### PR TITLE
fixes #9016 by implementing user auth

### LINKED ISSUE #9016: support user authentication for JMS connections
**Description:**
If I want to configure a JMS provider through JNDI, the current configuration parameters do not allow me to set an explicit username and password for the actual connection. The documentation uses ActiveMQ, which allows the credentials to be set through the connection URL, but this is not really applicable to other JMS providers.

Therefore I would really like to extends the JMS connection configuration, so it can hold credentials in addition the current properties.

I would provide a pull request for this issue, but since I'm just starting with ballerina, I am not sure what the correct place for such credentials is. Therefore I would suggest two solution which seam feasable in my opinion and ask for some feedback of what would be the right approach here.

Option 1:
```
jms:Connection conn = new({
    initialContextFactory:"com.tibco.tibjms.naming.TibjmsInitialContextFactory",
    providerUrl:"tcp://server1:7222,tcp://server2:7222",
    properties:{
        "java.naming.security.principal":"admin",
        "java.naming.security.credentials":"admin",
        "ConnectionUsername":"admin",
        "ConnectionPassword":"admin"
    }
});
```

Option 2:
```
jms:Connection conn = new({
    initialContextFactory:"com.tibco.tibjms.naming.TibjmsInitialContextFactory",
    providerUrl:"tcp://server1:7222,tcp://server2:7222",
    connectionUsername:"admin",
    connectionPassword:"admin",
    properties:{
        "java.naming.security.principal":"admin",
        "java.naming.security.credentials":"admin"
    }
});
```

**Steps to reproduce:**
Configuring a TIBCO EMS as JMS provider fails, since no authentication can be given.
```
jms:Connection conn = new({
    initialContextFactory:"com.tibco.tibjms.naming.TibjmsInitialContextFactory",
    providerUrl:"tcp://server1:7222,tcp://server2:7222",
    properties:{
        "java.naming.security.principal":"admin",
        "java.naming.security.credentials":"admin"
    }
});
```
**Affected Versions:**
v0.970.1

**Suggested Labels (optional):**
Type/New Feature, component/stdlib

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by JensWalter**
## Purpose
fixes #9016 by implementing user auth

### CODE REVIEW

File: stdlib/ballerina-jms/src/main/ballerina/jms/connection.bal [hunk 1]
```diff
@@ -46,11 +46,15 @@ documentation { Configurations related to a JMS connection
...
     F{{connectionFactoryName}} JMS connection factory to be used in creating JMS connections
+    F{{username}} Username for the JMS connection
+    F{{password}} Password for the JMS connection
     F{{properties}} Additional properties use in initializing the initial context
 }
 public type ConnectionConfiguration {
     string initialContextFactory = "wso2mbInitialContextFactory";
     string providerUrl = "amqp://admin:admin@ballerina/default?brokerlist='tcp://localhost:5672'";
     string connectionFactoryName = "ConnectionFactory";
+    string username;
```
Comment (VijithaEkanayake) [inline stdlib/ballerina-jms/src/main/ballerina/jms/connection.bal]: We need to define username and password as nullable strings as [1], if not you'll probably get an empty string value here[1] for username and password when the user doesn't configure username and password.

[1].
```
public type ConnectionConfiguration {
    string initialContextFactory = "wso2mbInitialContextFactory";
    string providerUrl = "amqp://admin:admin@ballerina/default?brokerlist='tcp://localhost:5672'";
    string connectionFactoryName = "ConnectionFactory";
    string username?;
    string password?;
    map properties;
};
```

[2]. https://github.com/ballerina-platform/ballerina-lang/pull/9212/files#diff-01450798131cea146644def4f81988c7R116

Comment (hasithaa) [reply stdlib/ballerina-jms/src/main/ballerina/jms/connection.bal]: @VijithaEkanayake, @shafreenAnfar Shall we merge this PR to separate branch and then fix said improvements. After that, we can merge this to master.

Comment (JensWalter) [reply stdlib/ballerina-jms/src/main/ballerina/jms/connection.bal]: I actually tried putting in the question marks, for the optional elements, but then the build failed with the following statement:
error: ballerina/jms/connection.bal:57:20: extraneous input '?'
error: ballerina/jms/connection.bal:58:20: extraneous input '?'

How would I get around this issue?

Comment (VijithaEkanayake) [reply stdlib/ballerina-jms/src/main/ballerina/jms/connection.bal]: @JensWalter Could you try as below. "?" should be associated with the type not with the variable. I've accidentally added it to the variable name in the sample. Apologies for the confusion made.

```
public type ConnectionConfiguration {
    string initialContextFactory = "wso2mbInitialContextFactory";
    string providerUrl = "amqp://admin:admin@ballerina/default?brokerlist='tcp://localhost:5672'";
    string connectionFactoryName = "ConnectionFactory";
    string? username;
    string? password;
    map properties;
};
```

Comment (JensWalter) [reply stdlib/ballerina-jms/src/main/ballerina/jms/connection.bal]: yes, that seems to work much better.
Now that I changed the definition of the type, does that have any influence on how ballerina maps values into the java space. As of right now, I'm only checking for 'null', should I also check for empty values?

current code section:
```            
String username = connectionConfig.getStringField(Constants.ALIAS_USERNAME);
String password = connectionConfig.getStringField(Constants.ALIAS_PASSWORD);
if (username != null && password != null) {
     return connectionFactory.createConnection(username, password);
} else {
     return connectionFactory.createConnection();
}
```

Comment (VijithaEkanayake) [reply stdlib/ballerina-jms/src/main/ballerina/jms/connection.bal]: Ballerina never assigns a null value for a data type unless the user specifically defines the type should be nullable. The Ballerina has set of default values for each of the data types. for example, default of string is empty string default of int is 0, boolean is false etc... Currently, we don't have a proper document to refer those values and I created issue[1] to track the documentation.

In here, As we define the username and password as nullable we don't need to check for the empty string. Checking the empty string is dependent on the use case. for example, If any case user has passed empty values for these parameters we might need to check the empty strings as well, assuming that empty string are not allowed to pass as parameter values.

[1]. https://github.com/ballerina-platform/ballerina-lang/issues/9226

Comment (JensWalter) [reply stdlib/ballerina-jms/src/main/ballerina/jms/connection.bal]: I made the `username` and `password` optional and changed the cheking for empty values from 'null' to `isEmpty()`

File: stdlib/ballerina-jms/src/main/java/org/ballerinalang/net/jms/JMSUtils.java [hunk 2]
```diff
@@ -99,10 +99,25 @@ public static Connection createConnection(Struct connectionConfig) {
...
+        }
+
         try {
             InitialContext initialContext = new InitialContext(properties);
             ConnectionFactory connectionFactory = (ConnectionFactory) initialContext.lookup(factoryName);
-            return connectionFactory.createConnection();
+            String username = connectionConfig.getStringField(Constants.ALIAS_USERNAME);
+            String password = connectionConfig.getStringField(Constants.ALIAS_PASSWORD);
+
+            if (!username.isEmpty() && !password.isEmpty()) {
```
Comment (VijithaEkanayake) [inline stdlib/ballerina-jms/src/main/java/org/ballerinalang/net/jms/JMSUtils.java]: As we have specified username and passwords are nullable fields in connection.bal [1]. Shall we also include the null check as well. Unless it'll lead to a NPE when the user doesn't provide a username or password. You can use method[2] perform this operation.

[1]. https://github.com/ballerina-platform/ballerina-lang/pull/9212/files#diff-1c8397e65ef4ec76bbff9c22853bdf0cR57
[2]. https://github.com/ballerina-platform/ballerina-lang/blob/master/stdlib/ballerina-jms/src/main/java/org/ballerinalang/net/jms/JMSUtils.java#L147

**Pull request merged**